src/analysis/asymptotics/negligible.lean

```diff
@@ -0,0 +1,224 @@
+/-
+Copyright (c) 2021 Devon Tuma. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Devon Tuma
+-/
+import analysis.asymptotics.asymptotics
+import analysis.asymptotics.specific_asymptotics
+import data.polynomial.eval
+
+/-!
+# Negligible Functions
+
+This file defines a predicate `negligible f` for a function satisfying
+  one of following equivalent definitions (The definition is in terms of the first condition):
+
+* `f` is `O(x ^ c)` for all (or equivalently sufficiently small) integers `c`
+* `f` is `O(p(x)⁻¹)` for all (or equivalently sufficiently large) polynomials `p`
+* `p(x) * f` is bounded for all polynomials `p`
+* `p(x) * f` tends to `𝓝 0` for all polynomials `p`
+
+-/
+
+namespace asymptotics
+
+open_locale topological_space
+open filter
+
+/-- A function `f` from an `ordered_comm_semiring` to a `normed_field` is negligible
+  iff `f(x)` is `O(x ^ c)` for all integers `c`. -/
+def negligible {α 𝕜 : Type*} [ordered_comm_semiring α] [normed_field 𝕜] [algebra α 𝕜]
+  (f : α → 𝕜) :=
+∀ (c : ℤ), is_O f (λ x, (algebra_map α 𝕜 x) ^ c) filter.at_top
```

Comment:
@hrmacbeth @sgouezel @semorrison @urkud  I rewrote the definition to allow any `ordered_comm_semiring` as the domain, assuming an algebra structure with the codomain. This definition ended up being fairly easy to work with, and most of the theorems about functions being negligible don't require any extra assumptions. This also had the added benefit of avoiding the need for ad hoc `coe` calls in a lot of the lemmas.

Some of the properties of negligible functions require additional assumptions about the algebra_map tending to at_top or the algebra_map eventually having norm greater than or equal to one.